### PR TITLE
client/libcephfs: Don't let one unmount screw up another mount

### PR DESCRIPTION
**Pull request by adamemerson**
Share threadpools between mounts. Don't shut them down until the last
user goes away, and make sure to start it back up when needed.

Fixes: https://tracker.ceph.com/issues/47294
For real this time.
Maybe.

### CODE REVIEW

**Comment by adamemerson** [comment]
https://pulpito.ceph.com/aemerson-2021-01-12_03:25:56-fs:volumes-wip-47294-distro-basic-smithi/

File: src/libcephfs.cc [hunk 1]
```diff
@@ -37,11 +37,34 @@
...
 
 static mode_t umask_cb(void *);
-ceph::async::io_context_pool icp;
+namespace {
+// Set things up this way so we don't start up threads until mount and
+// kill them off when the last mount goes away, but are tolerant to
+// multiple mounts of overlapping duration.
+std::shared_ptr<ceph::async::io_context_pool> get_icp(CephContext* cct)
+{
+  std::weak_ptr<ceph::async::io_context_pool> icwp;
```
Comment (batrick) [inline src/libcephfs.cc:46]: Is this supposed to be static? I'm confused because it looks like this method always creates a new io_context_pool?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/libcephfs.cc (hunk 1 above)
Comment (adamemerson) [reply src/libcephfs.cc]: Yes, thank you.

**Comment by adamemerson** [comment]
https://pulpito.ceph.com/aemerson-2021-01-12_22:32:47-fs:volumes-wip-47294-distro-basic-smithi/

A run with the update. Though that the other one worked suggests this was the root cause.

File: src/libcephfs.cc [hunk 2]
```diff
@@ -37,11 +37,34 @@
...
-ceph::async::io_context_pool icp;
+namespace {
+// Set things up this way so we don't start up threads until mount and
+// kill them off when the last mount goes away, but are tolerant to
+// multiple mounts of overlapping duration.
+std::shared_ptr<ceph::async::io_context_pool> get_icp(CephContext* cct)
+{
+  static std::weak_ptr<ceph::async::io_context_pool> icwp;
+
+  auto icp = icwp.lock();
```
Comment (batrick) [inline src/libcephfs.cc:48]: My understanding is that this is not thread-safe. The `weak_ptr` either needs to be protected by a lock or access needs to be atomic. We've done something like this elsewhere in Ceph: https://github.com/ceph/ceph/blob/6c805c640527f9600fe04369ac4a5d8ec8583707/src/common/ceph_context.h#L272-L279

This is available in C++20 but that doesn't help us now: https://en.cppreference.com/w/cpp/memory/weak_ptr/atomic2

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/libcephfs.cc [hunk 3]
```diff
@@ -37,11 +37,34 @@
...
+  if (icp)
+    return icp;
+
+  icp = std::make_shared<ceph::async::io_context_pool>();
+
+  auto icp2 = icwp.lock();
+  if (icp2)
+    return icp2;
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/libcephfs.cc]: Not if we're holding a lock.

File: src/libcephfs.cc (hunk 2 above)
Comment (adamemerson) [reply src/libcephfs.cc:52]: Fixed

**Comment by adamemerson** [comment]
http://pulpito.front.sepia.ceph.com:80/aemerson-2021-01-14_00:32:20-fs:volumes-wip-47294-distro-basic-smithi/

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
Thanks @adamemerson !

**Pull request merged**